### PR TITLE
feat: move to v4

### PR DESCRIPTION
**Pull request by avik-pal**
- [x] Fixes ManifoldProjection to use the actual formulation from Hairer 3
    - [x] Note that this is a breaking change (removing NonlinearSolve, users need to load it manually). Currently the version is kept as 3.10 since SciMLSensitivity has a dep on this causing a circular test dep, so can't test with v4.
    - [x] Also updates GeneralDomain as a consequence
- [x] Removes unused deps

### CODE REVIEW

File: src/manifold.jl [hunk 1]
```diff
@@ -33,30 +20,34 @@ properties.
 
 ## Arguments
 
-  - `g`: The residual function for the manifold.
```
Comment (avik-pal) [inline src/manifold.jl:36]: Similar to how our other packages work, I am making `manifold` st the IIP/OOP needs to align with the ODE problem

File: src/manifold.jl [hunk 2]
```diff
@@ -33,30 +20,34 @@ properties.
...
+    `g(u, p)`. Specify it as `Val(::Bool)` to disable runtime branching. If `nothing`,
+    we attempt to infer it from the signature of `g`.
+  - `residual_prototype`: The size of the manifold residual. If it is not specified,
+    we assume it to be same as `u` in the inplace case. Else we run a single evaluation of
+    `manifold` to determine the size.
   - `kwargs`: All other keyword arguments are passed to
-    [NonlinearSolve.jl](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/).
+    [NonlinearSolve.jl](https://docs.sciml.ai/NonlinearSolve/stable/basics/solve/) if
+    `nlsolve` is not `missing`.
+  - `autodiff`: The autodifferentiation algorithm to use to compute the Jacobian if
```
Comment (avik-pal) [inline src/manifold.jl:45]: making this mandatory for the time being. Once NonlinearSolveBase is released, we can reconsider that.

**Comment by avik-pal** [comment]
Needs https://github.com/SciML/LinearSolve.jl/pull/543 before we can test this

**Comment by avik-pal** [comment]
@ChrisRackauckas the runs pass. I am bumping the version to 4. This + https://github.com/SciML/SciMLSensitivity.jl/pull/1124 will drop one of the big deps in SciMLSensitivity

**Review by ChrisRackauckas — APPROVED** [review]

**Pull request merged**